### PR TITLE
fix(model): resolve Gemma 4 architecture bugs

### PR DESCRIPTION
**Pull request by cuichenx**
## What changed

This PR contains the Gemma 4 architecture/conversion bugs discovered while investigating #4610. The focused dense-softcap fix has merged separately in #4645.

1. **E2B heterogeneous FFNs**
   - preserves `use_double_wide_mlp`
   - allocates double-wide FC1 and FC2 only in the final shared-KV layers
   - keeps the existing optimized MCore MLP kernels; `Gemma4DenseMLP` only provides the correct per-layer config width
   - marks double-wide models as heterogeneous for distributed checkpointing; a real narrow/wide save/load roundtrip is covered in tests
2. **26B-A4B MoE state and runtime graph**
   - handles current MCore's `partial(MoELayer, ...)` builder
   - directly maps trainable router scale/projection, per-expert scale, and pre/post-FFN norms
   - removes the prior lossy BF16 router/shared-prenorm fusion and its inverse
   - runs Hugging Face's distinct router, shared-expert, and routed-expert inputs while retaining MCore expert dispatch, grouped experts, auxiliary loss, and MoE recompute
   - preserves MCore's TE checkpoint path for FP8/FP4 whole-MoE recompute
   - shares the raw K projection with V before applying their independent HF normalizations on `attention_k_eq_v` layers
3. **31B dense attention config**
   - maps `sliding_window` into MCore's inclusive window tuple
   - preserves `attention_k_eq_v` for full-attention layers
4. **Training and execution safety**
   - marks new replicated norm/router parameters for sequence-parallel gradient synchronization
   - changes the functional SFT smoke to TP=2
   - rejects MCore orchestration modes bypassed by the custom MoE forward instead of silently accepting incorrect behavior
5. **Complete config export fidelity**
   - reconstructs official dense and MoE `Gemma4Config` nesting, while retaining flat `Gemma4TextConfig` for text mode
   - preserves dtype, layer pattern, local/global RoPE, head geometry, shared-KV/double-wide/window/K=V fields, PLE fields, MoE count/top-k/width, vision/audio configs, special tokens, and official model types
6. **Parity tooling**
   - removes the parity script's manual second softcap now that the model output layer applies it

Related to #4610. The issue-closing final-logit-softcap fix is merged via #4645.

## Why the fusion code is removed

The old implementation approximated Hugging Face's three-input MoE graph by folding router scale divided by pre-FFN norm-2 into the router projection and folding pre-FFN norm-1 divided by norm-2 into shared-expert weights. That workaround was lossy in BF16, omitted 120 trainable tensors from 26B export, and did not reproduce the training graph. The replacement stores and applies the original parameters directly; the code was removed rather than relocated.

## Validation environment

- `nvcr.io/nvidia/nemo:26.06`, Python 3.12, Transformers 5.8.1
- base `45186600a`; MCore `0823c731ed7d793aef047b6a64f2dbbf32bf6e2c`
- 2x RTX 6000 Ada 48 GiB
- roundtrip seed 1234; parity seed 4610; deterministic BF16 eval, TF32 disabled

Pinned official checkpoints:

- `google/gemma-4-E2B-it@70af34e20bd4b7a91f0de6b22675850c43922a03`
- `google/gemma-4-E4B-it@fee6332c1abaafb77f6f9624236c63aa2f1d0187`
- `google/gemma-4-26B-A4B-it@20da991ab4afab98e8f910c4a2e8f4fbefc404ad`
- `google/gemma-4-31B-it@3548789868c5356dbf307c98e6f609007b82b3eb`

### HF -> MCore -> HF state correspondence

Full `GEMMA4_CONVERSION_MODE=auto` / `Gemma4VLBridge` roundtrips (the final review fixes do not change weight mappings):

| Variant | TP/EP | Exact tensors | Compared elements | Missing/extra | Dtype mismatch | Error |
|---|---:|---:|---:|---:|---:|---:|
| E2B | 1/1 | 2,011/2,011 | 5,123,178,979 | 0/0 | 0 | exactly 0 |
| E4B | 1/1 | 2,130/2,130 | 7,996,157,418 | 0/0 | 0 | exactly 0 |
| 26B-A4B | 1/2 | 1,013/1,013 | 25,805,936,206 | 0/0 | 0 | exactly 0 |
| 31B | 2/1 | 1,188/1,188 | 31,273,088,876 | 0/0 | 0 | exactly 0 |

The separate dense text/LLM path is also bitwise exact for every language tensor: E2B 600 tensors / 4,647,449,891 elements, E4B 719 / 7,518,069,034, and 31B 832 / 30,697,345,340. Its omitted source keys are exactly the expected vision/audio/projector tensors.

### HF logit and generation parity

These end-to-end results were measured with this architecture fix combined with #4645's final softcap. Both pre-softcap and final post-softcap logits were captured for short/long prompts and an 8x output-head saturation control. All short/long final argmaxes agree, and the saturation cases prove exactly one cap.

| Variant | Short post max/mean abs | Long post max/mean abs | 8x post max/mean abs | Greedy new IDs |
|---|---:|---:|---:|---:|
| E2B | 0.375 / 0.005325 | 0.5 / 0.1183 | 0 / 0 | 8/8 exact |
| E4B | 0.5 / 0.05957 | 0.625 / 0.2076 | 2.5625 / 0.000411 | 8/8 exact |
| 26B-A4B | 2.1875 / 0.2029 | 7.84375 / 0.7178 | 3.25 / 2.00e-5 | 8/8 exact |
| 31B | 0.6875 / 0.1157 | 0.828125 / 0.06829 | 2.4375 / 0.006614 | 8/8 exact |

The corrected 26B K=V normalization changes greedy agreement from the first 3/8 tokens to exact 8/8. Residual BF16 fused-kernel drift later moves a few close router decisions and is amplified by sparse routing; the post-softcap logits are not claimed to be exact.

### Config and distributed-checkpoint fidelity

- all seven supported official Auto/VL and text combinations reconstruct with zero architecture-field mismatch and exact vision/audio config dictionaries
- a real four-layer E2B model with narrow and double-wide FFNs saves to MCore distributed checkpoint, loads into a new model, and matches every tensor exactly
- behavioral tests distinguish HF's `VNorm(raw_K)` from the incorrect `VNorm(KNorm(raw_K))`
- FP8 and FP4 tests require TE checkpoint dispatch and fail if the ordinary tensor-parallel checkpoint path is selected

### Focused tests and quality gates

- focused Gemma 4 bridge/model/provider/VL suite: `311 passed`
- TP=2 + sequence-parallel, 2-layer/4-expert Gemma 4 VL SFT: 10/10 iterations, 80 samples, grad norm 39.938, zero skipped/NaN iterations; validation and test completed
- `uv run pre-commit run --all-files`
- `git diff --check`

Exact state roundtrip is necessary but not sufficient: it does not exercise model construction, graph placement, router APIs, nonlinear output config, gradient synchronization, or generation. The construction, forward/parity, TP=2 training, and saturation checks above cover those separate failure classes.

## Deliberately out of scope

To keep this bug-fix PR reviewable, the separable audio-input plumbing, heterogeneous FLOP reporting, and selectable Transformer Engine RMSNorm backend/recipe defaults are not included. They remain preserved in the development backup for focused follow-ups.

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

**Comment by tushtush-mc** [comment]
@cuichenx

There appears to be an issue when converting a checkpoint from Megatron to Hugging Face format.

In `run_config.yaml`, `window_size` is saved as a list:

```yaml
window_size:
  - 511
  - 0
```

However, Transformers expects `window_size` to be an integer. Changing it to the following resolves the issue:

```yaml
window_size: 511
```

### Reproduction

```bash
python convert_checkpoints.py export \
  --hf-model "google/gemma-4-E4B-it" \
  --megatron-path "path/to/gemma-4-E4B-it_megatron" \
  --hf-path "path/to/gemma-4-E4B-it_megatron_back_to_hf"
```

Screenshot:

Error encountered during Megatron-to-Hugging-Face conversion

**Comment by cuichenx** [comment]
@tushtush-mc thanks for reporting this. fixing in #4800
